### PR TITLE
Polarion report generation for release-readiness and send across email report to the team.

### PR DESCRIPTION
**Pull request by neha-gangadhar**
Polarion report generation for release-readiness and send across email report to the team.

Steps:
1. Extracts required data from polarion portal both tier-wise and component-wise.
2. Generates percentage of automation completion.
3. Generates delta of automated status from last run/week.
4. Generates no. of automated tests in not approved status. 
4. Updates all the data in separate sheets of a main google spreadsheet.
5. Creates bar chart based on the values and uploads to google sheet.
6. Generates email report and sends across to the team.

Signed-off-by: Neha Gangadhar <ngangadh@redhat.com>

### CODE REVIEW

File: tools/release-readiness/polarion/export_comp_data_to_sheets.py [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+                print("start_c inner for loop:", start_c)
+            start_c = self.get_next_col(start_c)
+            print("start_c outer for loop:", start_c)
+        start_c = self.get_prev_col(start_c)
+        print("start_c outside loop:", start_c)
+        ws.format("A" + row_count + ":" + start_c + row_count, self.format_cell())
+
+        """
+        #Charts part addition needed.
+        if row_count == "3":
```
Comment (sunilkumarn417) [inline tools/release-readiness/polarion/export_comp_data_to_sheets.py:137]: can we take this doc string out, if not needed.

File: tools/release-readiness/polarion/export_to_sheets.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
+import datetime
+from re import S
+from time import sleep
+
+import google_sheet_chart
+from Chart import Chart_json
+from gspread_formatting import *
+
+
+class export_data:
```
Comment (sunilkumarn417) [inline tools/release-readiness/polarion/export_to_sheets.py:10]: same comment as above.

File: tools/release-readiness/polarion/export_to_sheets.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+    """Module to export data to google sheet"""
+
+    import gspread
```
Comment (sunilkumarn417) [inline tools/release-readiness/polarion/export_to_sheets.py:13]: same comment as above

File: tools/release-readiness/polarion/google_sheet_chart.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    "https://www.googleapis.com/auth/drive",
+]
+
+READONLY_SCOPES = [
+    "https://www.googleapis.com/auth/spreadsheets.readonly",
+    "https://www.googleapis.com/auth/drive.readonly",
+]
+
+
+class Make_charts:
```
Comment (sunilkumarn417) [inline tools/release-readiness/polarion/google_sheet_chart.py:18]: same as above comment.

File: tools/release-readiness/polarion/html_template/all_status_component_wise_data.html [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            <TD style="color: blue"><a href="https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitems?query=project.id%3ACEPH%20AND%20casecomponent.KEY%3A{{each_key}}%20AND%20tier.KEY%3A(tier1%20tier2%20tier3%20tier4)%20AND%20NOT%20status%3A(approved%20inactive)%20AND%20caseautomation.KEY%3Aautomated)">{{each_key}}</a></TD>
+            <TD>{{comp_aut_delta[each_key]}}</TD>
+        {% endif %}
+    </TR>
+    {% endfor %}
+
+</table>
+
+</body>
+</html>
```
Comment (sunilkumarn417) [inline tools/release-readiness/polarion/html_template/all_status_component_wise_data.html:49]: please add a new line.

File: tools/release-readiness/polarion/html_template/automation_delta.html [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                    <TD style ="background-color: #ffff00">{{tier4_percent}}%</TD>
+         {% elif tier4_percent > 75 %}
+                <TD style ="background-color: #00FF7F">{{tier4_percent}}%</TD>
+         {% endif %}
+
+    </TR>
+</table>
+
+</body>
+</html>
```
Comment (sunilkumarn417) [inline tools/release-readiness/polarion/html_template/automation_delta.html:96]: please add a newline.

File: tools/release-readiness/polarion/html_template/automation_delta_component.html [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+		<TD style="color: blue">{{each_key}}</TD>
+          {% for each_ind in range(4) %}
+             <TD style ="background-color:powderblue;">{{aut_delta[each_key][each_ind]}}</TD>
+          {% endfor %}
+        </TR>
+    {% endfor %}
+
+
+</body>
+</html>
```
Comment (sunilkumarn417) [inline tools/release-readiness/polarion/html_template/automation_delta_component.html:52]: same comment as above.

File: tools/release-readiness/polarion/html_template/component_wise_data.html [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        <TD style="background-color: #ff7f8a"> <50% </TD>
+        <TD>N/A</TD>
+        </TR>
+        <TR> ↑ The increase from last week
+        <br> ↓ The decrease from last week
+        <br> ~ No change
+        </TR>
+
+</body>
+</html>
```
Comment (sunilkumarn417) [inline tools/release-readiness/polarion/html_template/component_wise_data.html:94]: same comment as above.

**Review by sunilkumarn417 — COMMENTED** [review]
Hi @neha-gangadhar ,

This is a good Start.
I think we can discuss few things over a call.

File: tools/release-readiness/polarion/Chart.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
+from gspread import utils
+
+
+class Chart_json:
```
[the remark this answers is not in the document]
Comment (neha-gangadhar) [reply tools/release-readiness/polarion/Chart.py:4]: done

File: tools/release-readiness/polarion/Chart.py [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+        self._domain = domain
+        self._range = range
+        self._anchorCell = anchorCell
+        self._sheetId = sheetId
+        self._request = dict()
+        self._chart_sheetId = chart_sheetId
+
+    def Create_json(self, id):
+        """Function to create chart"""
+        self._request["addChart"] = dict()
```
[the remark this answers is not in the document]
Comment (neha-gangadhar) [reply tools/release-readiness/polarion/Chart.py:27]: Not taking up in this PR. Would make a note for next one.

File: tools/release-readiness/polarion/Chart.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+        )
+        anchell_body["sheetId"] = json_body["sheetId"]
+        anchell_body["rowIndex"] = json_body["endRowIndex"]
+        anchell_body["columnIndex"] = json_body["endColumnIndex"]
+        self._request["addChart"]["chart"]["position"]["overlayPosition"][
+            "anchorCell"
+        ] = anchell_body
+        return self._request
+
+    def Update_json(self, chart_id):
```
[the remark this answers is not in the document]
Comment (neha-gangadhar) [reply tools/release-readiness/polarion/Chart.py:130]: done

File: tools/release-readiness/polarion/Chart.py [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+        """Function to update chart"""
+        self._request["updateChartSpec"] = dict()
```
[the remark this answers is not in the document]
Comment (neha-gangadhar) [reply tools/release-readiness/polarion/Chart.py:132]: Not taking up in this PR. Would make a note for next one.

File: tools/release-readiness/polarion/Chart.py [hunk 13]
```diff
@@ -0,0 +1,237 @@
+from gspread import utils
+
+
+class Chart_json:
+    """Represents a chart in a sheet."""
+
+    def __init__(self, title, domain, range, anchorCell, sheetId, chart_sheetId):
+        """
```
[the remark this answers is not in the document]
Comment (neha-gangadhar) [reply tools/release-readiness/polarion/Chart.py:8]: done

File: tools/release-readiness/polarion/export_comp_data_to_sheets.py [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+import time
+from re import S
+from time import sleep
+
+import google_sheet_chart
+from Chart import Chart_json
+from gspread_formatting import *
+
+
+class export_data:
```
[the remark this answers is not in the document]
Comment (neha-gangadhar) [reply tools/release-readiness/polarion/export_comp_data_to_sheets.py]: done

File: tools/release-readiness/polarion/export_comp_data_to_sheets.py (hunk 1 above)
Comment (neha-gangadhar) [reply tools/release-readiness/polarion/export_comp_data_to_sheets.py]: done

File: tools/release-readiness/polarion/export_to_sheets.py (hunk 2 above)
Comment (neha-gangadhar) [reply tools/release-readiness/polarion/export_to_sheets.py]: done

File: tools/release-readiness/polarion/html_template/all_status_component_wise_data.html (hunk 5 above)
Comment (neha-gangadhar) [reply tools/release-readiness/polarion/html_template/all_status_component_wise_data.html]: done

File: tools/release-readiness/polarion/html_template/component_wise_data.html (hunk 8 above)
Comment (neha-gangadhar) [reply tools/release-readiness/polarion/html_template/component_wise_data.html]: done

**Comment by neha-gangadhar** [comment]
hi @sunilkumarn417 

PR comments addressed, please review and help merge.

Thanks,
Neha

**Comment by neha-gangadhar** [comment]
hi @psathyan 

could you please help review the PR.

Thanks,
Neha

**Pull request closed without merging**